bin/crew

```diff
@@ -556,77 +561,114 @@ def download
+
+    case File.basename(filename)
+    # Sources that download with curl
+    when /\.zip$/i, /\.(tar(\.(gz|bz2|xz|lz))?|tgz|tbz|txz)$/i, /\.deb$/i
+      # Recall file from cache if requeseted
+      if CREW_IS_CACHING
+        cachefile = CREW_CACHE_DIR + filename
+        if File.file?(cachefile)
+          if Digest::SHA256.hexdigest( File.read(cachefile) ) == sha256sum then
+            FileUtils.cp cachefile, CREW_BREW_DIR, verbose: @fileutils_verbose
+            puts "Archive found in cache".lightgreen
+            return {source: source, filename: filename}
           else
-          puts 'Cached archive checksum mismatch. :/ Will download.'.lightred
+            puts 'Cached archive checksum mismatch. :/ Will download.'.lightred
+          end
+        puts 'Cannot find cached archive. :/ Will download.'.lightred
         end
       end
-    end
-
-    system "#{CURL} --retry 3 -#{@verbose}#LC - --insecure \'#{url}\' --output #{filename}"
-
-    abort 'Checksum mismatch. :/ Try again.'.lightred unless
-      Digest::SHA256.hexdigest( File.read(filename) ) == sha256sum
-    puts "Archive downloaded".lightgreen
+      # Download file if not cached
+      system "#{CURL} --retry 3 -#{@verbose}#LC - --insecure \'#{url}\' --output #{filename}"
+      abort 'Checksum mismatch. :/ Try again.'.lightred unless
+        Digest::SHA256.hexdigest( File.read(filename) ) == sha256sum
+      puts "Archive downloaded".lightgreen
+      # Stow file in cache if requested
+      if CREW_IS_CACHING
+        puts 'Caching downloaded archive'.lightgreen
+        FileUtils.cp filename, CREW_CACHE_DIR, verbose: @fileutils_verbose
+      end
 
-    if ENV['CREW_CACHE_OPT']
-      puts 'Caching downloaded archive'.lightgreen
-      FileUtils.cp filename, CREW_CACHE_DIR, verbose: @fileutils_verbose
+    # Sources that download with git
+    when /\.git$/i
+      # Recall repository from cache if requested
+      if CREW_IS_CACHING
+        cachefile = CREW_CACHE_DIR + '/' + filename + '.tar.xz'
+        if File.file?(cachefile)
+          if system "sha256sum -c #{cachefile}.sha256"
+            FileUtils.mkdir @extract_dir
+            system "tar x#{@verbose}f #{cachefile} -C #{@extract_dir}"
+            return {source: source, filename: filename}
+          else
+            puts 'Cached repository checksum mismatch. :/ Will download.'.lightred
+          end
+        else
+          puts 'Cannot find cached repository. :/ Will download.'.lightred
+        end
+      end
+      # Download file if not needed
+      Dir.mkdir @extract_dir
+      Dir.chdir @extract_dir do
+        system 'git init'
+        system 'git config advice.detachedHead false'
+        system 'git config init.defaultBranch master'
+        system "git remote add origin #{@pkg.source_url}", exception: true
+        if @pkg.git_hash
+          system "git fetch origin"
+          system "git checkout #{@pkg.git_hash}", exception: true
```

Comment:
Do we even need this line if the same thing happens on line 624?

Reply:
It's not the same thing, we do need both lines.

Reply:
Why would you need to checkout twice?  That doesn't make any sense.  Maybe line 624 should be replaced with line 620?

Reply:
What if `git_hash` is empty?

Reply:
How about this logic?  Trying to simplify...
```
if @pkg.git_branch.is_a?(Hash)
  system "git clone #{@pkg.source_url}"
else
  system "git clone --depth 1 #{@pkg.source_url}"
end
Dir.chdir File.basename("#{@pkg.source_url}", ".git") do
  system "git checkout #{@pkg.git_branch}"
end
```
@pkg.git_branch could be a branch, tag or hash.  Easier?

Reply:
What if someone needs to specify a branch/tag on a specific branch?

Reply:
Well, I guess you do have a point to that argument.  The point I'm trying to make is `git clone` can replace all that `git init`, `git fetch`, `git remote add`, etc. extra unnecessary stuff.

Reply:
@satmandu what do you think about that? You were the one who suggested the git init and git config sequence.

Reply:
@uberhacker git clone is simpler, but the git init, fetch, etc nonsense is **insanely powerful** because then you don't need ANY other logic to just fetch a git hash - and we really don't want to do git fetches of branches since those are moving targets, right?

Reply:
That is true, it would be better to only fetch tags and hashes.

Reply:
Also, that could also be this in a couple of  lines,  which is really quick:

```
Dir.chdir @extract_dir do
        system 'git init'
        system 'git config advice.detachedHead false'
        system 'git config init.defaultBranch master'
        system "git fetch --depth 1 origin #{@pkg.git_hashtag}"
        system 'git checkout FETCH_HEAD'
        puts "Repository downloaded".lightgreen
```
I would actually lean against officially supporting a branch at all... You want to support commits... otherwise, you have no idea what you are downloading.

Reply:
you can't clone a hash or tag like that, only a branch, at least from the testing I've done.

Reply:
Alright, this is some really good feedback.  It looks messy but after your explanation, it makes more sense.

Reply:
I will let @satmandu make the executive decision to merge this or not.

Reply:
working on this...